### DISCUSSION TITLE
How is Microsoft paint implemented?

### DISCUSSION THREAD

**Discussion by ClearSeve** [microsoft/microsoft-ui-xaml#8536] [General]
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/26833671/1c76378f-60e9-4439-a728-f9e3e4d8fa4c)
I want to implement paint tools similar to this, including windows for HWND and windows for XAML. How is this implemented?
Isn't this using Islands?

**Comment by abart27** [comment]
Concluding from the handle tree, it utilizes XAML Islands for interoperability of Win32 with UWP (+ WinUI2)

Presumably the legacy Win32 drawing canvas has been put into a XAML Island, as it is both visually and functionally identical to the old one.

**Comment by mdtauk** [comment]
> Concluding from the handle tree, it utilizes XAML Islands for interoperability of Win32 with UWP (+ WinUI2)
> 
> Presumably the drawing canvas has been put into a XAML Island, as it is both visually and functionally identical to the old one.

I would guess the drawing surface retains the Win32 code, and the Window and the rest of the UI is now in XAML.

I am curious how the "ribbon" is being done.  The flyouts are XAML but WinUI 2 or 3 doesn't yet have a Ribbon control, so is that also Win32 spawning Xaml flyouts?  Or is there an Internal version of the Ribbon that may be released for everyone at some point, like the File Explorer team did with their CollectionView controls

**Comment by ClearSeve** [comment]
I want to create a program that combines WinUI3 and Win32 hwnd. There is a technology called "Islands", but WinUI3 does not support "Islands" yet, so I am curious about how Microsoft does this, and the same goes for Notepad

**Comment by abart27** [comment]
> The flyouts are XAML but WinUI 2 or 3

The entire XAML part of the app is UWP + WinUI2, as indicated by the `CoreWindow` in Spy++

**Comment by mdtauk** [comment]
> > The flyouts are XAML but WinUI 2 or 3
> 
> The entire XAML part of the app is UWP + WinUI2, as indicated by the `CoreWindow` in Spy++

So the drawing area is still a Win32 component?  But the rest of it is WinUI 2.  So the ribbon is not a skinned MFC one, and is in fact an internal CustomControl?
